drivers/mpu9250/MPU9250_mag.cpp

```diff
@@ -0,0 +1,462 @@
+{
+	int retVal = 0;
+	uint8_t b = 0;
+
+	// Configure a write operation to the mag using Slave 4.
+	retVal = write_imu_reg_verified(MPUREG_I2C_SLV4_ADDR,
+			MPU9250_AK8963_I2C_ADDR, 0xff);
+	if (retVal != 0) {
+		return retVal;
+	}
+
+	// Set the mag register address to write to using Slave 4.
+	retVal = write_imu_reg_verified(MPUREG_I2C_SLV4_REG, reg, 0xff);
+	if (retVal != 0) {
+		return retVal;
+	}
+
+	// Set the value to write in the I2C_SLV4_DO register.
+	retVal = write_imu_reg_verified(MPUREG_I2C_SLV4_DO, val, 0xff);
+	if (retVal != 0) {
+		return retVal;
+	}
+
+	// Read the current value of the Slave 4 control register.
+	retVal = _imu.readReg(MPUREG_I2C_SLV4_CTRL, b);
+	if (retVal != 0) {
+		return retVal;
+	}
+
+	// Set I2C_SLV4_EN bit in I2C_SL4_CTRL register without overwriting other
+	// bits.
+	b |= BITS_I2C_SLV4_EN;
+	// Trigger the data transfer from the byte now stored in the SLV4_DO register.
+	retVal = _imu.writeReg(MPUREG_I2C_SLV4_CTRL, b);
+	if (retVal != 0) {
+		return retVal;
+	}
+
+	// Continuously check I2C_MST_STATUS regsiter value for the completion
+	// of I2C transfer until timeout
+	retVal = _imu.readReg(MPUREG_I2C_MST_STATUS, b);
+	if (retVal != 0) {
+		return retVal;
+	}
+
+	int loop_ctrl = 1000; // wait up to 1000 * 1ms for completion
+	while (((b & BITS_I2C_SLV4_DONE) == 0x00) && (--loop_ctrl)) {
+		usleep(1000);
+		retVal = _imu.readReg(MPUREG_I2C_MST_STATUS, b);
+		if (retVal != 0) {
+			return retVal;
+		}
+	}
+
+	if (loop_ctrl == 0) {
+		DF_LOG_ERR("I2C transfer to mag timed out");
+		return -1;
+	}
+
+#ifdef MPU9250_MAG_DEBUG
+	DF_LOG_INFO("Magnetometer register %u set to %u", reg, val);
+#endif
+
+	return 0;
+}
+
+int MPU9250_mag::process(struct mag_data &data)
+{
+#ifdef MPU9250_MAG_DEBUG
+	static int hofl_bit_counter = 0;
+#endif
+
+	// Check magnetic sensor overflow HOFL bit set.  No need to check the data ready bit, since
+	// the sample rate divider should provide new samples at the correct interval.
+	if (data.mag_st2 & BIT_MAG_HOFL) {
+#ifdef MPU9250_MAG_DEBUG
+		if ((++hofl_bit_counter % 1000) == 0) {
+			DF_LOG_ERR("overflow HOFL bit set (x1000)");
+		}
+#endif
+		return MAG_ERROR_DATA_OVERFLOW;
+	}
+
+	data.mag_x = ImuSensor::swap16(data.mag_x);
+	data.mag_y = ImuSensor::swap16(data.mag_y);
+	data.mag_z = ImuSensor::swap16(data.mag_z);
+
+	// H_adj = H * ((ASA-128)*0.5/128 + 1)
+	//       = H * ((ASA-128) / 256 + 1)
+	// H is the raw compass reading, ((ASA-128) / 256 + 1) has been
+	// computed and stored in compass_cal_f:
+	// _mag_sens_adj[i] = (float) (((float) asa[i] - 128.0) / 256.0) + 1.0f;
+	data.mag_x = (int16_t) ((int) data.mag_x * _mag_sens_adj[0]);
+	data.mag_y = (int16_t) ((int) data.mag_y * _mag_sens_adj[1]);
+	data.mag_z = (int16_t) ((int) data.mag_z * _mag_sens_adj[2]);
+
+	// Convert the native units of the sensor (micro-Teslas) to Gauss
+	data.mag_x /= 100;
+	data.mag_y /= 100;
+	data.mag_z /= 100;
+
+	// Swap magnetometer x and y axis
```

Comment:
Is the mag inside the MPU9250 in a different orientation than the accel/gyro? Or am I missing something?